testing/endtoend/components/eth1/depositor.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v3/config/params"
+	contracts "github.com/prysmaticlabs/prysm/v3/contracts/deposit"
+	eth "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	e2e "github.com/prysmaticlabs/prysm/v3/testing/endtoend/params"
+	"github.com/prysmaticlabs/prysm/v3/testing/endtoend/types"
+	"github.com/prysmaticlabs/prysm/v3/testing/util"
+)
+
+const depositGasLimit = 4000000
+
+var gweiPerEth = big.NewInt(int64(params.BeaconConfig().GweiPerEth))
+
+func amtInGwei(deposit *eth.Deposit) *big.Int {
+	amt := big.NewInt(0).SetUint64(deposit.Data.Amount)
+	return amt.Mul(amt, gweiPerEth)
+}
+
+func computeDeposits(offset, nvals int, partial bool) ([]*eth.Deposit, error) {
+	balances := make([]uint64, offset+nvals)
+	partialIndex := len(balances) // set beyond loop invariant so by default nothing gets partial
+	if partial {
+		// Validators in this range will get 2 half (MAX_EFFECTIVE_BALANCE/2) deposits.
+		// Upper half of the range of desired validator indices is used because these can be easily
+		// duplicated with a slice from this offset to the end.
+		partialIndex = offset + nvals/2
+	}
+	// Start setting values at `offset`. Lower elements will be discarded - setting them to zero
+	// ensures they don't slip through and add extra deposits.
+	for i := offset; i < len(balances); i++ {
+		if i >= partialIndex {
+			// these deposits will be duplicated, resulting in 2 deposits summing to MAX_EFFECTIVE_BALANCE
+			balances[i] = params.BeaconConfig().MaxEffectiveBalance / 2
+		} else {
+			balances[i] = params.BeaconConfig().MaxEffectiveBalance
+		}
+	}
+	deposits, _, err := util.DepositsWithBalance(balances)
+	if err != nil {
+		return []*eth.Deposit{}, err
+	}
+
+	// if partial = false, these will be a no-op (partialIndex == len(deposits)),
+	// otherwise it will duplicate the partial deposits.
+	deposits = append(deposits[offset:], deposits[partialIndex:]...)
+
+	return deposits, nil
+}
+
+type Depositor struct {
+	// The EmptyComponent type is embedded in the Depositor so that it satisfies the ComponentRunner interface.
+	// This allows other components or e2e set up code to block until its Start method has been called.
+	types.EmptyComponent
+	Key       *keystore.Key
+	Client    *ethclient.Client
+	ChainID   *big.Int
+	NetworkId *big.Int
+	cd        *contracts.DepositContract
+	sent      *DepositHistory
+}
+var _ types.ComponentRunner = &Depositor{}
+
+// History exposes the DepositHistory value for a Depositor.
+func (d *Depositor) History() *DepositHistory {
+	if d.sent == nil {
+		d.sent = &DepositHistory{}
+	}
+	return d.sent
+}
+
+// DepositHistory is a type used by Depositor to keep track of batches of deposit for test assertion purposes.
+type DepositHistory struct {
+	byBatch map[types.DepositBatch][]*SentDeposit
```

Comment:
Consider wrapping map access around some kind of mutex to make this threadsafe

Reply:
This isn't possible with current e2e but I can add it for future proofing.

Reply:
updated, ptal